8.0 connector flow

codingforfun opened the pull request:
Add 'connector_flow' module. This module provides a minimal framework for common
data import/export tasks based on the OpenERP Connector.

Commit:
first stable release of connector_flow

Commit:
add external_dependencies in travis.yml

Commit:
resolve duplicate xml id

Commit:
link to tutorial

codingforfun committed:
Port 'connector_flow' module to new v8 API.

codingforfun committed:
Remove obsolete module 'fix_bug_1316948'. This bug has been fixed in odoo core.

codingforfun committed:
Remove obsolete function.

codingforfun committed:
fix pep8 errors

OSguard commented on connector_flow/wizard/run_task.py:
```diff
@@ -0,0 +1,81 @@
+        #   There is no convention or clear specification that the return of
+        #   do_run() always is an impexp.file id. In fact I know no real
+        #   example there this is the case
+        #
```
@tremlin can you say something to that?

guewen commented on connector_flow/__openerp__.py:
```diff
@@ -0,0 +1,76 @@
+{
+    'name': 'Connector-based task flow for import/export',
+    'version': '0.1',
+    'description': """This module provides a minimal framework for common
```
Could you extract the description in a README.rst file based in this template: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst ?

guewen commented on connector_flow/__openerp__.py:
```diff
@@ -0,0 +1,76 @@
+A very brief tutorial can be found at http://blog.initos.com/?p=1220
+""",
+    'category': 'Connector',
+    'author': 'initOS GmbH & Co. KG',
```
Could you add `Odoo Community Association (OCA)` as author?
`'author': 'initOS GmbH & Co. KG,Odoo Community Association (OCA)'`

guewen commented on connector_flow/chunk.py:
```diff
@@ -0,0 +1,40 @@
+class imp_exp_chunk(models.Model):
+    _name = 'impexp.chunk'
+    _description = 'Structured (parsed) data from a file' +\
+                   ' to be imported/exported'
```
You can remove `+\` and write:

```
    _description = ('Structured (parsed) data from a file'
                    ' to be imported/exported')
```

guewen commented on connector_flow/chunk.py:
```diff
@@ -0,0 +1,40 @@
+from openerp import models, fields, api
+
+
+class imp_exp_chunk(models.Model):
```
CamelCase is preferred for new API classes

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+    return session.pool.get('impexp.task') \
+        .related_action(session.cr, session.uid,
+                        job=job, context=session.context)
+
```
You should be able to write:

```
session.env['impexp.task'].related_action(job=job)
```

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+@related_action(action=impexp_related_action)
+def run_task(session, model_name, ids, **kwargs):
+    return session.pool.get('impexp.task') \
+        .run_task(session.cr, session.uid, ids, **kwargs)
```
You should be able to write:

```

session.env['impexp.task'].run_task(**kwargs)

```

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+            if 1 < flow_start_count:
+                raise exceptions. \
+                    ValidationError('The start of a task flow '
+                                    'has to be unique')
```
Should be translatable

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+                           'max_retries': self.max_retries})
+        else:
+            method = run_task
+        result = method(ConnectorSession(self.env.cr,
```
Could be:

```
ConnectorSession.from_env(self.env)
```

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+        start_task = False
+        for task in flow.task_ids:
+            if task.flow_start:
+                start_task = task
```
Missing a `break`?

guewen replied:
Am I wrong?

codingforfun replied:
No you'r not. I already changed this, but it isn't detected by GH somehow. It's just the line below your comment. Please have a look at the full diff.

guewen replied:
Yes, sure. Thanks for the change

guewen commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,161 @@
+
+    @api.model
+    def related_action(self, job=None, **kwargs):
+        task_instance = self.browse(job.args[1]).get_task_instance()
```
Error if `job` is None. Should be checked before trying to access `job.args`.

guewen commented on connector_flow/task/abstract_task.py:
```diff
@@ -0,0 +1,121 @@
+
+    def __init__(self, cr, uid, ids):
+        self.session = ConnectorSession(cr, uid)
+        self._id = ids[0]
```
Maybe add an `assert len(ids) == 1` before?

guewen commented on connector_flow/task/abstract_task.py:
```diff
@@ -0,0 +1,121 @@
+        return retval
+
+    def create_file(self, filename, data):
+        ir_attachment_id = self.session.env['ir.attachment'].\
```
Rename `ir_attachment_id` to `ir_attachment` because it is a record

guewen commented on connector_flow/task/abstract_task.py:
```diff
@@ -0,0 +1,121 @@
+            create({'name': filename,
+                    'datas': b64encode(data),
+                    'datas_fname': filename})
+        file_id = self.session.env['impexp.file'].\
```
Rename `file_id` to `impexp_file` because it is a record (and `file` is a builtin)

guewen commented on connector_flow/task/abstract_task.py:
```diff
@@ -0,0 +1,121 @@
+class abstract_chunk_read_task(abstract_task):
+    """Task that reads (and processes) an existing chunk of data"""
+
+    def run(self, chunk_id=None, **kwargs):
```
Return if chunk_id is None?

guewen commented on connector_flow/task/abstract_task.py:
```diff
@@ -0,0 +1,121 @@
+    """Task that writes (and feeds) data as a chunk"""
+    def write_and_run_chunk(self, chunk_data, chunk_name,
+                            async=True, **kwargs):
+        chunk_id = self.session.env['impexp.chunk'].\
```
Rename `chunk_id` to `chunk`, it's a record

guewen commented on connector_flow/wizard/run_task.py:
```diff
@@ -0,0 +1,81 @@
+        task_id = False
+        for task in self.flow_id.task_ids:
+            if task.flow_start:
+                task_id = task.id
```
Missing `break`?

guewen commented on connector_flow_example_ftp/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+{
+    'name': 'Example (Partner Export via FTP) for connector_flow',
+    'version': '0.1',
+    'description': """This module provides a demo task flow for the
```
Could you extract the description in README.rst based on this template: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

guewen commented on connector_flow_example_ftp/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+(name,zip code).""",
+    'category': 'Connector',
+    'license': 'AGPL-3',
+    'author': 'initOS GmbH & Co. KG',
```
Could you add 'Odoo Community Association (OCA)'?

guewen commented on connector_flow_example_ftp/improved_ftp.py:
```diff
@@ -0,0 +1,52 @@
+    _columns = {
+        'task': fields.selection(_get_available_tasks, string='Task',
+                                 required=True),
+    }
```
Should be migrated to the new API as well

guewen replied:
I was a bit laconic in my comment. I think it would be better to have the example modules using the new API because they will serve as example for new modules and so they should be as good as possible. Do you agree with that?

guewen commented on connector_flow_example_ftp/res_partner.py:
```diff
@@ -0,0 +1,56 @@
+        #  as list of lists, corresponding to the structured rows
+        #  in the export file
+        result_list = [['Name', 'ZIP Code']]
+        partner_obj = self.session.pool.get('res.partner')
```
Use `session.env`

guewen commented on connector_flow_example_product/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+    'version': '0.1',
+    'description': """This module provides a demo task flow for the
+*connector_flow* module. It implements a very simple product catalog import
+from CSV""",
```
Same remark than the others: README.rst and OCA author

guewen commented on connector_flow_example_product/product.py:
```diff
@@ -0,0 +1,56 @@
+    _columns = {
+        'task': fields.selection(_get_available_tasks, string='Task',
+                                 required=True),
+    }
```
Could you migrate it to the new API?

guewen commented on connector_flow_example_product/product.py:
```diff
@@ -0,0 +1,56 @@
+    def _get_available_tasks(self, cr, uid, context=None):
+        return super(product_catalog_import_task, self) \
+            ._get_available_tasks(cr, uid, context=context) \
+            + [('product_catalog_import', 'Produkt Catalog Import')]
```
s/Produkt/Product/

guewen commented on connector_flow_xls/__openerp__.py:
```diff
@@ -0,0 +1,52 @@
+    'name': 'XLS Import for connector_flow',
+    'version': '0.1',
+    'description': """This module provides an XLS import task for the
+ connector_flow module""",
```
Same remark than the others

guewen commented on connector_flow_xls/task/xls_import.py:
```diff
@@ -0,0 +1,54 @@
+    _columns = {
+        'task': fields.selection(_get_available_tasks, string='Task',
+                                 required=True),
+    }
```
New API?

codingforfun committed:
Update description

codingforfun committed:
Codestyle

codingforfun committed:
Change to v8 API.

codingforfun committed:
Make error messages translatable.

codingforfun committed:
Check for valid job argument.

codingforfun committed:
Check for single id

codingforfun committed:
Typo

codingforfun committed:
Add description readme files

Viggor commented on connector_flow/task/ftp_download.py:
```diff
@@ -0,0 +1,111 @@
+            ('ftp_download', 'FTP Download')]
+
+    def ftp_download_class(self):
+        return FtpDownload
```
You can use the module in https://github.com/OCA/knowledge/pull/45 to do this task (download/upload). I reuse a very similar logic to do it but this feature (file transmition) should be done with another module for a better flexibility.

guewen replied:
Since this is a former addon, my feeling is that you @Viggor should make a proposal to integrate your new module in/with this one.

Viggor replied:
My module do only file import/export, so it's a base for other module which want to handle files. If too many things are in the same module it's more hard to be flexible and use just the feature you need. So I think the connector_flow should be based on my module for import/export and keep the others features in the connector_flow module.

bealdav replied:
@Viggor invest a lot of time to build a proposal for exchange of the files. It's based inspired by connector_file from c2c, connector from initOS and file_exchange from akretion. We are ok to drop our solution 'file-exchange', but please have a look at the work of @Viggor https://github.com/OCA/knowledge/pull/45/files.
@OSguard , @tremlin , @codingforfun. Do you want than @Viggor make a PR on your PR to make the 2 projects full complementary of your real good work ?
We need to be the more as possible in OCA. Then you can use @viggor modules (which reuse the better idea of everybody), and when somebody want more to manage data in files, use connector_flow modules. Don't you think ?

OSguard replied:
i think this is a great idea and we should combine the work. So we mark this PR as "work in progress" and you make a new PR to our branch and their we can discuss the best solution to combine our work.

Viggor replied:
Ok nice, i'll do it.
But why you make your PR in this repo and not in the Knowledge repo as you suggest here https://github.com/OCA/connector-interfaces/issues/5 ?

Viggor replied:
PR done here https://github.com/initOS/connector-interfaces/pull/1

Viggor commented on connector_flow/task/ftp_upload.py:
```diff
@@ -0,0 +1,80 @@
+            ('ftp_upload', 'FTP Upload')]
+
+    def ftp_upload_class(self):
+        return FtpUpload
```
Same comment about download.

codingforfun committed:
Fix typo from refactoring.

codingforfun committed:
Move task flow configuration from demo to data section.

Always install task flow data. Since this is an example module it should be complete and functional after install.

codingforfun committed:
Do not overwrite user configuration after install.

lmignon commented on connector_flow/README.rst:
```diff
@@ -0,0 +1,60 @@
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
```
{project_repot} {module_name} and {version} are place holders and must be replaced with the appropriate values. 
- s/{project_repot}/connetor-interface 
- s/{version}/8.0
- s/{module_name}/connector_flow

lmignon commented on connector_flow/__openerp__.py:
```diff
@@ -0,0 +1,55 @@
+    'application': False,
+    'images': [
+    ],
+    'css': [
```
you can remove empty entries (demo, css, js, qweb)

lmignon commented on connector_flow/impexp_task.py:
```diff
@@ -0,0 +1,158 @@
+    transitions_in_ids = fields.One2many('impexp.task.transition',
+                                         'task_to_id',
+                                         string='Incoming Transitions')
+    flow_start = fields.Boolean(string='Start of a Task Flow')
```
may be a select=True since the filed is used in the domain defined at line 96

codingforfun committed:
Change to v8 API.

codingforfun committed:
Move task flow configuration from demo to data section.

codingforfun committed:
Change to v8 API.

lmignon commented on connector_flow/task/ftp_download.py:
```diff
@@ -0,0 +1,111 @@
+from openerp import models, api
+
+from .abstract_task import AbstractTask
+import ftputil
```
you should enclose imports for ftputil in a try catch clause to avoid problems when the module is not installed and thus the ftplutil is not required in the python path.

``` python
try:
    import ftputil
    import ftputil.session
except ImportError:
    pass
```

codingforfun replied:
ftputil is mentioned in the manifest as an external dependency of the module. So it's needed anyway. 
However I'm thinking of slicing the module further to contain only the basic task flow chain and moving extensions like FTP to extra modules.

lmignon replied:
@codingforfun Since the dependency is declared in the manifest,  Odoo will prevent the installation if it's not present in the pythonpath. But don't forget that even if your module is not installed, it will be imported into the runtime and thus the server won't start if the ftputil lib is not installed. That's why you should protect the import in a try catch to allow the server to start without the lib installed if your module is not installed.

codingforfun replied:
Ok. I wasn't aware of this. Are you really sure? I tried this and added an import statement to an not existing python module into an uninstalled odoo module and the server was starting just fine on my local install. Does this depend on some setting?

lmignon replied:
@codingforfun 
As I remember that was the case previously but I can't reproduce the error now.

guewen replied:
@codingforfun I think it may happen for instance when you do an "Update Module List" and maybe for other reasons. I don't know exacly when but it does happen.

codingforfun replied:
I tried this also, because I also added my `not_existing_python_module` also as an external python dependency to the manifest to model the exact same case as with `ftputil`. 
I worked without problems. Trying to install the modul produced the expected error message due to the misseng external dependency.

lmignon replied:
@codingforfun @guewen I'm sure that previously we had to protect the import to avoid errors at startup in such a case. I can't reproduce the error now and I'm very curious to know what has changed in Odoo. In fact we need to know what has changed since this 'enclosing' has been used in other addons and could be removed if it's no more required.

lmignon commented on connector_flow/task/ftp_upload.py:
```diff
@@ -0,0 +1,80 @@
+from openerp import models, api
+from .abstract_task import AbstractTask
+from base64 import b64decode
+import ftputil
```
see previous comment to protect import on external dependencies

lmignon commented on connector_flow_example_ftp/README.rst:
```diff
@@ -0,0 +1,43 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/{project_repo}/issues/new?body=module:%20{module_name}%0Aversion:%20{version}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
replace place holders

lmignon commented on connector_flow_example_ftp/__openerp__.py:
```diff
@@ -0,0 +1,48 @@
+    'application': False,
+    'images': [
+    ],
+    'css': [
```
remove empty entries

lmignon commented on connector_flow_example_product/README.rst:
```diff
@@ -0,0 +1,42 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/{project_repo}/issues/new?body=module:%20{module_name}%0Aversion:%20{version}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
replace place holders

lmignon commented on connector_flow_example_product/__openerp__.py:
```diff
@@ -0,0 +1,49 @@
+    'application': False,
+    'images': [
+    ],
+    'css': [
```
remove empty entries

lmignon commented on connector_flow_xls/README.rst:
```diff
@@ -0,0 +1,41 @@
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/{project_repo}/issues/new?body=module:%20{module_name}%0Aversion:%20{version}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
replace place holders

lmignon commented on connector_flow_xls/__openerp__.py:
```diff
@@ -0,0 +1,50 @@
+    'application': False,
+    'images': [
+    ],
+    'css': [
```
remove empty entries

lmignon commented on connector_flow_xls/task/xls_import.py:
```diff
@@ -0,0 +1,54 @@
+from openerp.osv import orm, fields
+from openerp.addons.connector_flow.task.csv_import import table_row_import
+
+from xlrd import open_workbook
```
``` python
try:
    from xlrd .....
except ImportError:
    pass
```

codingforfun committed:
Fix placeholders

codingforfun committed:
Remove empty manifest entries

codingforfun committed:
Remove obsolete code fragment.

codingforfun committed:
Remove 'required' from impexp.file to impexp.task relation.

Making the relation to the task mandatory makes uninstalling tasks impossible while keeping their results.

codingforfun committed:
Simplify: Remove output pane from wizard

codingforfun committed:
Fix flake8 errors

Commit:
Merge remote-tracking branch 'upstream/8.0' into 8.0-connector_flow

Conflicts:
	.travis.yml

Commit:
[FIX] fix lint issues

paulius-sladkevicius commented on connector_flow/chunk_view.xml:
```diff
@@ -0,0 +1,27 @@
+            <field name="model">impexp.chunk</field>
+            <field name="type">form</field>
+            <field name="arch" type="xml">
+                <form string="Task Form" version="7.0" create='false'>
```
version="7.0" is not needed anymore

paulius-sladkevicius commented on connector_flow/task/csv_import.py:
```diff
@@ -0,0 +1,90 @@
+            chunk_id = self.session.create('impexp.chunk',
+                                           {'name': name,
+                                            'data': simplejson.dumps(data),
+                                            'file_id': file.id})
```
Getting warning: 
`openerp.addons.connector_flow.task.csv_import: Deprecated: run at line 62: 'Session(...).create(...)' has been deprecated in favor of 'Session(...).env['model'].create(...)`

rousseldenis commented on connector_flow/file_view.xml:
```diff
@@ -0,0 +1,33 @@
+                  action="file_view_action"/>
+
+    </data>
+</openerp>
```
Maybe should you add some filters to improve search ?

paulius-sladkevicius committed:
connector_flow: remove deprecated connector API

paulius-sladkevicius committed:
connector_flow: remove deprecated view version tag

Commit:
Merge pull request #7 from HBEE/8.0-connector_flow-view

connector_flow: remove deprecated view version tag

Commit:
Merge pull request #6 from HBEE/8.0-connector_flow-connector-api

connector_flow: remove deprecated connector API

rousseldenis commented:
@codingforfun Can you do the remaining changes that should make this PR ok ?
Otherwise :+1: for me. Using in production

codingforfun commented:
What's the most important leftovers in your opinion. 
Unfourtunately I don't had much time for this recently, especially to follow all the discussion in the related PRs. I will check what's possible.

pedrobaeza commented:
Is there still interest in this?
